Fix BaseHtmlElement

lippserd opened the pull request:
Hi Tom,

I fixed some bugs in `BaseHtmlElement` and added PHPDoc where necessary. I added test cases and commit descriptions which should explain the changes. But these are the highlights:

* `render()` did not respect a possible `getDefaultAttributes()` override
* Elements which did not specify a tag were rendered as `<>$content`. It's now just `$content`
* It was allowed to create void elements with content. The content was silently ignored upon rendering. We now throw an Exception here
* `getTag()` overrides were not respected in some places
* `in_array()` is replaced with `isset()` for void detection and the cache has been removed. `setTag()` did not invalidate the cache which may have lead to false results when calling `isVoid()` or `wantsClosingTag()`
* `setVoid()` explicitly supports `null`. Was working before but now it is documented

Cheers,
Eric

lippserd committed:
Respect getDefaultAttributes() overrides in render()

Subclasses of BaseHtmlElement may override the property
defaultAttributes or the method getDefaultAttributes() in order to
specify the default attributes. Both ways must be supported when
rendering the attributes. Before, getDefaultAttributes() overrides had
no effect. So, renderAttributes() has been removed in favor of calling
getAttributes()->render().

lippserd committed:
Remove BaseHtmlElement::canBeRendered()

This method is not used. Further, there already is
Html::canBeRenderedAsString() for that purpose.

lippserd committed:
Throw exception if void elements have content

Our focus is to provide valid and safe HTML5 wherever we are in control.

lippserd committed:
Add PHPDoc to $voidElements

lippserd committed:
Use getTag() instead of $tag

getTag() may be overriden.

lippserd committed:
Don't cache isVoidElement()

in_array is replaced with isset which makes the cache no longer
necessary. This automatically fixes a bug in the previous implementation
where setTag() did not invalidate the cache.

lippserd committed:
Rename isVoidElement() to isVoid()

isVoid() says it all.

lippserd committed:
Move getTag() before setTag()

Getters come first.

lippserd committed:
Add more PHPDoc to BaseHtmlElement

lippserd committed:
Support null in setVoid()

null resets void detection to its default behavior.

lippserd commented:
Could we just ignore coding standards for test cases?

lippserd committed:
Use $name instead of $key in setAttribute()

The Attributes class uses $name as well.

lippserd commented:
One more thing: Should we enforce a tag in `wrap()` or does wrap work without a tag too?

Thomas-Gelf commented:
I'd strongly suggest to enforce a tag in an HtmlElement, there is no element without a tag in HTML. There is HtmlDocument for everything else, the distinction is intentional. We could of course have one class for "all the things", but that feels weird. How would you render an HtmlElement with attributes but no tag?

Thomas-Gelf committed:
BaseHtmlElementTest: helper classes are fine

lippserd commented:
Sounds good to me. The only thing which comes to my mind then is to check whether a tag is set in `render()` and `isVoid()` and if not, throw an exception. We can't to that in `getTag()` because it may be overridden.

Thomas-Gelf commented:
That's fine, please go for it and remove the single commit in question. And don't forget to pull first, I pushed a "fix" for your helper classes.

lippserd commented:
Affirmative.

lippserd committed:
Assert tag in getTag()

HTML elements must have tag. getTag() is now final and throws an
exception if the element does not have tag. In order to dynamically
provide a tag, the procted method tag() is introduced.

Thomas-Gelf committed:
BaseHtmlElement: fix setVoid(null), related test

Thomas-Gelf committed:
BaseHtmlElement: RuntimeException is fine

Thomas-Gelf reviewed (approved):
Nice, thank you!
